### PR TITLE
[PyTorch] C&W Implementation

### PR DESCRIPTION
**Pull request by iamgroot42**
Ported from TF version (tried to keep it as close as possible).
I noticed that there was another PR recently requested for a merge on the same attack [here](https://github.com/tensorflow/cleverhans/pull/1141). However, there are some subtle differences (in favour of this version):

- Uses tensors to keep track of lower, upper bounds, etc (instead of pythonic lists)
- Retains option for early stopping
- Infer number of classes from shape of logits (instead of making it a parameter)
- Retain option to log useful information

@carlini it would be great if you could have a look at it!

### CODE REVIEW

File: cleverhans/future/torch/tests/test_attacks.py [hunk 1]
```diff
@@ -27,13 +29,23 @@ def forward(self, x):
...
+
+
 class CommonAttackProperties(CleverHansTest):
 
   def setUp(self):
     super(CommonAttackProperties, self).setUp()
     self.model = SimpleModel()
     self.x = torch.randn(100, 2)
-    self.normalized_x = torch.rand(100, 2) # truncated between [0, 1)
+    self.normalized_x = torch.rand(100, 2)  # truncated between [0, 1)
```
Comment (iamgroot42) [inline cleverhans/future/torch/tests/test_attacks.py:48]: This change (and similar ones below it that are seemingly not relevant) were auto generated when I ran autopep8. Nonetheless, these changes should be here anyway

**Comment by michaelshiyu** [comment]
Hi @iamgroot42, please respect other's work. There is already a ready-to-be-reviewed PR doing exactly the same thing. And the differences you mentioned between yours and that one are minor at best. Duplicate work also adds unnecessary workload to the moderators/reviewers. 

Even if you do spot important performance issues in some PR, it would be much better if you open a discussion with the project maintainers as well as the developer(s) of the other PR and either contribute yourself to that PR or have the author(s) modify their PR. Opening a duplicate PR without discussing with the project maintainers or the author(s) of the existing PR is really not the most efficient or respectful thing to do. If this was ok, anyone could basically copy all existing PRs and open new ones with minor tweaks and nobody would be able to do any work. Thanks.

**Comment by iamgroot42** [comment]
@michaelshiyu  fair enough. Closing this PR. Let's switch over to yours.

P.S. There was no active open PR (yours opened in January but remained unmerged for a long time) when I started working on mine. Perhaps I should've checked in between to track status :)

**Comment by michaelshiyu** [comment]
Thanks @iamgroot42, I appreciate it. That PR has been opened in Jan. but I've been actively updating it, incorporating suggestions from reviews, and clearing merge conflicts. And it is now waiting for another review from the maintainers. 

Thanks again and feel free to make changes there.
